README.md
# laravel5-otp

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

The code was taken directly from [https://github.com/teckwei1993/laravel-otp](https://github.com/teckwei1993/laravel-otp) and tweaked to our need. Please use original repo for your work.

A package for Laravel One Time Password (OTP) generator and validation without Eloquent Model, since it done by *Cache*.
The cache connection same as your laravel cache config and it supported: "apc", "array", "database", "file", "memcached", "redis"

## Install

Via Composer

``` bash
$ composer require achinthailabs/laravel5-otp
```

## Add Service Provider & Facade

**For Laravel 5.5+**

Once the package is added, the service provider and facade will be auto discovered.

**For Laravel 5.2 / 5.3 / 5.4**

Add the ServiceProvider to the providers array in `config/app.php`:

```php
Achinthailabs\Otp\OtpServiceProvider::class
```

Add the Facade to the aliases array in `config/app.php`:

```php
'Otp' => Achinthailabs\Otp\OtpFacade::clas
```

## Configuration

Publish config and language file

```bash
php artisan vendor:publish --provider="Achinthailabs\Otp\OtpServiceProvider"
```

This package publishes an `otp.php` file inside your applications's config folder which contains the settings for this package. Most of the variables are bound to environment variables, you may add Key-Value pair to the `.env` file in the Laravel application.

```
OTP_FORMAT=numeric
OTP_LENGTH=6
OTP_SENSITIVE=false
OTP_EXPIRES_TIME=15
OTP_ATTEMPT_TIMES=5
OTP_REPEATED=true
OTP_DEMO=false
```

## Usage

### Generate OTP

```php
Otp::generate(string $identifier)
```

* `$identifier`: The identity that will be tied to the OTP.

#### Sample

```php
use OTP;

// in controller

$password = Otp::generate('reg:name@domain.com');
```

This will generate a OTP that will be valid for 15 minutes.

### Validate OTP

```php
Otp::validate(string $identifier, string $password)
```

* `$identifier`: The identity that is tied to the OTP.
* `$password`: The password tied to the identity.

#### Sample

```php
use OTP;

// in controller

$result = Otp::validate('reg:name@domain.com', '123456');
```

#### Responses

**On Success**

```object
{
  "status": true
}
```

**Invalid OTP**

```object
{
  "status": false,
  "error": "invalid"
}
```

**Expired**

```object
{
  "status": false,
  "error": "expired"
}
```

**Max attempt**

```object
{
  "status": false,
  "error": "max_attempt"
}
```

* Reached the maximum allowed attempts, default 10 times with each identifier

### Validate OTP by Laravel Validation

```php
// in a `FormRequest`

use Achinthailabs\Otp\Rules\OtpValidate;

public function rules()
{
    return [
        'code' => ['required', new OtpValidate('change-email:name@domain.com')]
    ];
}

// in a controller

$request->validate([
    'code' => ['required', new OtpValidate('change-email:name@domain.com')]
]);
```

### Validate OTP by session id

```php
// Otp class

$result = Otp::validate('123456');

// in a `FormRequest`

use Achinthailabs\Otp\Rules\OtpValidate;

public function rules()
{
    return [
        'code' => ['required', new OtpValidate()]
    ];
}

// in a controller

$request->validate([
    'code' => ['required', new OtpValidate()]
]);
```

* The setting without identifier will automatically use the session ID as the default, and the OTP generation and verification will be completed in same session (browser's cookies).

## Advanced Usage

### Generate OTP with options

```php
$password = Otp::setLength(8)->setFormat('string')->setExpires(60)->setRepeated(false)->generate('identifier-key-here');

// or array option

$password = Otp::generate('identifier-key-here', [
    'length' => 8,
    'format' => 'string',
    'expires' => 60,
    'repeated' => false
]);
```

* `setLength($length)`: The length of the password. Default: 6
* `setFormat($format)`: The format option allows you to decide which generator implementation to be used when generating new passwords. Options: 'string','numeric','numeric-no-zero','customize'. Default: "numeric"
* `setExpires($minutes)`: The expiry time of the password in minutes. Default: 15
* `setRepeated($boolean)`: The repeated of the password. The previous password is valid when new password generated until either one password used or itself expired. Default: true

### Generate OTP with customize password

```php
$password = Otp::setCustomize('12345678ABC@#$')->generate('identifier-key-here');
```

* `setCustomize($string)`: Random letter from the customize string

### Validate OTP with specific attempt times

```php
$password = Otp::setAttempts(3)->validate('identifier-key-here', 'password-here');
```

* `setAttempts($times)`: The number of incorrect password attempts. Default: 5

### Validate OTP with case sensitive

```php
$password = Otp::setSensitive(true)->generate('identifier-key-here');

// validate

$result = Otp::setSensitive(true)->validate('identifier-key-here', 'password-here');

// in controller

use Achinthailabs\Otp\Rules\OtpValidate;

$request->validate([
    'code' => ['required', new OtpValidate('identifier-key-here', ['sensitive' => true])]
]);
```

* `setSensitive($boolean)`: Requiring correct input of uppercase and lowercase letters. Default: true

### Generate OTP with seperate password

```php
$password = Otp::setLength([4,3,4])->setSeparator(':')->generate('identifier-key-here');
```
**Sample password**

```
3526:126:3697
```

* `setLength($array)`: The length of the password, use array to separate each length.
* `setSeparator($string)`: The separator of the password. Default: "-"

### Validate OTP with extra data

```php
$password = Otp::setData(['user_id' => auth()->id()])->generate('login-confirmation');
```

* `setData($var)`: Allows you to get the extra data of OTP.

```php
// validate

$result = Otp::setDisposable(false)->validate('login-confirmation', 'password-here');

// in controller

use Achinthailabs\Otp\Rules\OtpValidate;

$request->validate([
    'code' => ['required', new OtpValidate('login-confirmation', ['disposable' => false])]
]);
```

* `setDisposable($boolean)`: The disposable of the Otp identifier, the different password is not valid when same identifier password used. Default: true

**On Success Response**

```object
{
  "status": true,
  "data": [
    "user_id": 10
  ]
}
```

* When you set disposable to `false`, you are able support different password with different extra data for different user in the same identifier key of the OTP.

### Validate OTP with skip using

```php
// validate

$result = Otp::setSkip(true)->validate('identifier-key-here', 'password-here');

// in controller

use Achinthailabs\Otp\Rules\OtpValidate;

$request->validate([
    'code' => ['required', new OtpValidate('identifier-key-here', ['skip' => true])]
]);
```

* `setSkip($boolean)`: Skip using the password when validate, which means you can reuse the password again. Default: false
* When there is an error response to the form request, it will skip using the password, but remember to `OTP::validate(...)`  in controller.

### Delete OTP

```php
Otp::forget('identifier-key-here');
```

* Delete all password with this specific identifier

### Delete specific password

```php
Otp::forget('identifier-key-here', 'password-here');
```

### Reset attempt times

```php
Otp::resetAttempt('identifier-key-here');
```

### Demo password

Add the following Key-Value pair to the `.env` file in the Laravel application.

```
OTP_DEMO=true
```

* Demo mode for development purposes, no need to use real password to validate.
* Default demo password: "1234", "123456", "12345678"
* 
## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email achintha@incubatelabs.com instead of using the issue tracker.

## Credits

- [Achintha Rodrigo][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/achinthailabs/laravel5-otp.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/achinthailabs/laravel5-otp/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/achinthailabs/laravel5-otp.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/achinthailabs/laravel5-otp.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/achinthailabs/laravel5-otp.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/achinthailabs/laravel5-otp
[link-travis]: https://travis-ci.org/achinthailabs/laravel5-otp
[link-scrutinizer]: https://scrutinizer-ci.com/g/achinthailabs/laravel5-otp/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/achinthailabs/laravel5-otp
[link-downloads]: https://packagist.org/packages/achinthailabs/laravel5-otp
[link-author]: https://github.com/achinthailabs
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `laravel5-otp` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](https://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/ExampleTest.php
<?php

declare(strict_types=1);

namespace achinthailabs\laravel5-otp;

class ExampleTest extends \PHPUnit\Framework\TestCase
{
    /**
     * Test that true does in fact equal true
     */
    public function testTrueIsTrue()
    {
        $this->assertTrue(true);
    }
}


src/Otp.php
<?php

namespace Achinthailabs\Otp;

use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Facade;
use Illuminate\Support\Str;

class Otp
{
    private $format = 'numeric';
    private $customize;
    private $length = 6;
    private $separator = '-';
    private $sensitive = false;
    private $expires = 15;
    private $attempts = 5;
    private $repeated = true;
    private $disposable = true;
    private $prefix = 'OTPPX_';
    private $data;
    private $skip = false;
    private $demo = false;
    private $demo_passwords = ['1234','123456','12345678'];

    public function __construct()
    {
        foreach(['format','customize','length','separator','sensitive','expires','attempts','repeated','disposable','prefix','data','demo','demo_passwords'] as $value){
            if(!empty(config('otp.'.$value))) $this->{$value} = config('otp.'.$value);
        }
    }

    public function __call(string $method, $params)
    {
        if(substr($method, 0, 3) != 'set'){
            return;
        }

        $property = Str::camel(substr($method, 3));
        if(! property_exists($this, $property)){
            return;
        }

        $this->{$property} = $params[0] ?? null;
        if($property == 'customize'){
            $this->format = 'customize';
        }
        return $this;
    }

    public function generate(string $identifier = null, array $options = [])
    {
        if(!empty($options)) foreach(['format','customize','length','separator','sensitive','expires','repeated','prefix','data'] as $value){
            if(isset($options[$value])) $this->{$value} = $options[$value];
        }

        if($identifier === null) $identifier = session()->getId();
        $array = $this->repeated ? $this->readData($identifier, []) : [];
        $password = $this->generateNewPassword();
        if(!$this->sensitive) $password = strtoupper($password);
        $array[md5($password)] = [
            'expires' => time()+$this->expires*60,
            'data' => $this->data
        ];
        $this->writeData($identifier, $array);
        return $password;
    }

    public function validate(string $identifier = null, string $password = null, array $options = [])
    {
        if(!empty($options)) foreach(['attempts','sensitive','disposable','skip'] as $value){
            if(isset($options[$value])) $this->{$value} = $options[$value];
        }

        if($password === null){
            if($identifier === null){
                throw new \Exception("Validate parameter can not be null");
            }
            $password = $identifier;
            $identifier = null;
        }

        if($this->demo && in_array($password,$this->demo_passwords)){
            return (object) [
                'status' => true,
                'demo' => true
            ];
        }

        if($identifier === null) $identifier = session()->getId();
        $attempt = $this->readData('_attempt_'.$identifier, 0);
        if($attempt >= $this->attempts){
            return (object) [
                'status' => false,
                'error' => 'max_attempt',
            ];
        }

        $codes = $this->readData($identifier, []);
        if(!$this->sensitive) $password = strtoupper($password);

        if(!isset($codes[md5($password)])){
            $this->writeData('_attempt_'.$identifier, $attempt+1);
            return (object) [
                'status' => false,
                'error' => 'invalid',
            ];
        }

        if(time() > $codes[md5($password)]['expires']){
            $this->writeData('_attempt_'.$identifier, $attempt+1);
            return (object) [
                'status' => false,
                'error' => 'expired',
            ];
        }

        $response = [
            'status' => true,
        ];

        if(!empty($codes[md5($password)]['data'])){
            $response['data'] = $codes[md5($password)]['data'];
        }

        if(!$this->skip) $this->forget($identifier, !$this->disposable ? $password : null);
        $this->resetAttempt($identifier);

        return (object) $response;
    }

    public function forget(string $identifier = null, string $password = null)
    {
        if($identifier === null) $identifier = session()->getId();
        if($password !== null){
            $codes = $this->readData($identifier, []);
            if(!isset($codes[md5($password)])){
                return false;
            }
            unset($codes[md5($password)]);
            $this->writeData($identifier, $codes);
            return true;
        }

        $this->deleteData($identifier);
        return true;
    }

    public function resetAttempt(string $identifier = null)
    {
        if($identifier === null) $identifier = session()->getId();
        $this->deleteData('_attempt_'.$identifier);
        return true;
    }

    private function generateNewPassword()
    {
        try{
            $formats = [
                'string' => $this->sensitive ? '23456789abcdefghjkmnpqrstuvwxyzABCDEFGHJKMNPQRSTUVWXYZ' : '23456789ABCDEFGHJKLMNPQRSTUVWXYZ',
                'numeric' => '0123456789',
                'numeric-no-zero' => '123456789',
                'customize' => $this->customize
            ];

            $lengths = is_array($this->length) ? $this->length : [$this->length];

            $password = [];
            foreach($lengths as $length){
                $password[] = substr(str_shuffle(str_repeat($x=$formats[$this->format], ceil($length/strlen($x)) )),1, $length);
            }

            return implode($this->separator, $password);
        }catch(\Exception $e){
            throw new \Exception("Fail to generate password, please check the format is correct.");
        }
    }

    private function writeData(string $key, $value)
    {
        $expires = $this->expires;
        if(app()->version() >= "5.8"){
            $expires *= 60;
        }
        return Cache::put($this->prefix.$key, $value, $expires*3);
    }

    private function readData(string $key, $default = null)
    {
        return Cache::get($this->prefix.$key, $default);
    }

    private function deleteData(string $key)
    {
        return Cache::forget($this->prefix.$key);
    }
}


src/Rules/OtpValidate.php
<?php

namespace Achinthailabs\Otp\Rules;

use Illuminate\Contracts\Validation\Rule;
use Achinthailabs\Otp\OtpFacade as Otp;

class OtpValidate implements Rule
{
    protected $identifier;
    protected $options;
    protected $attribute;
    protected $error;

    public function __construct(string $identifier = null, array $options = [])
    {
        $this->identifier = $identifier ?: session()->getId();
        $this->options = $options;
    }

    public function passes($attribute, $value): bool
    {
        $result = Otp::validate($this->identifier, $value, $this->options);
        if($result->status !== true){
            $this->attribute = $attribute;
            $this->error = $result->error;
            return false;
        }

        return true;
    }

    public function message(): string
    {
        return __('otp::messages.'.$this->error, [
            'attribute' => $this->attribute
        ]);
    }
}


src/OtpServiceProvider.php
<?php

namespace Achinthailabs\Otp;

use Illuminate\Support\ServiceProvider;

class OtpServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->alias(Otp::class, 'otp');
    }

    /**
     * Bootstrap any application services.
     *
     * @return void
     */
    public function boot()
    {
        if(app()->version() < "9.0"){
            $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'otp');

            $this->publishes([
                __DIR__.'/../resources/lang' => resource_path('lang/vendor/otp'),
                __DIR__.'/../config/otp.php' => config_path('otp.php')
            ]);
            return;
        }

        $this->loadTranslationsFrom(__DIR__.'/../lang', 'otp');

        $this->publishes([
            __DIR__.'/../resources/lang' => $this->app->langPath('vendor/otp'),
            __DIR__.'/../config/otp.php' => config_path('otp.php')
        ]);
    }
}


config/otp.php
<?php

return [

    /*
     * The format option allows you to decide
     * which generator implementation to be used when
     * generating new passwords.
     *
     * Here are the options:
     *  - string
     *  - numeric
     *  - numeric-no-zero
     *  - customize
     */

    'format' => env('OTP_FORMAT', 'numeric'),

    /*
    * The customize option required when option is customize.
    */

    'customize' => '123456789ABCDEFG@#$%',

    /*
     * The length of the password.
     */

    'length' => env('OTP_LENGTH', 6),

    /*
     * The separator of the password.
     */

    'separator' => '-',

    /*
     * Requiring correct input of uppercase and lowercase letters.
     */

    'sensitive' => env('OTP_SENSITIVE', false),

    /*
     * The expiry time of the password in minutes.
     */

    'expires' => env('OTP_EXPIRES_TIME', 15),

    /*
     * The number of incorrect password attempts.
     */

    'attempts' => env('OTP_ATTEMPT_TIMES', 5),

    /*
     * The repeated of the password.
     * The previous password is valid when new password generated
     * until either one password used or itself expired.
     */

    'repeated' => env('OTP_REPEATED', true),

    /*
     * The disposable of the Otp identifier.
     * The different password is not valid when same identifier password used.
     */

    'disposable' => true,

    /*
     * The prefix of the cache key to be used to store.
     */

    'prefix' => 'OTPPX_',

    /*
     * Demo mode for development purposes, no need to use real password to validate.
     */

    'demo' => env('OTP_DEMO', false),
    'demo_passwords' => ['1234','123456','12345678']

];
